test/Player.spec.ts

```diff
@@ -0,0 +1,287 @@
+import 'phaser';
+
+import { Player } from '../src/gameobjects/Player';
+
+import * as chai from 'chai';
+import { stub } from 'sinon';
+import * as sinonChai from 'sinon-chai';
+import { PlayerScene } from './scenes/PlayerScene';
+import { Block } from '../src/gameobjects/Block';
+import { Game, GameObjects } from 'phaser';
+
+chai.use(sinonChai);
+const expect = chai.expect;
+
+class TestGame extends Phaser.Game {
+    constructor(config: Phaser.Types.Core.GameConfig) {
+        super(config);
+    }
+}
+
+function dispatchKeyEvent(keyEvent: string, keyCode: any) {
+    window.dispatchEvent(
+        new KeyboardEvent(keyEvent, {
+            // @ts-ignore https://github.com/photonstorm/phaser/issues/2542
+            keyCode: keyCode,
+        })
+    );
+}
+
+function dispatchKeyDown(keyCode) {
+    dispatchKeyEvent('keydown', keyCode);
+}
+
+function dispatchKeyUp(keyCode) {
+    dispatchKeyEvent('keyup', keyCode);
+}
+
+describe('Player', () => {
+    var game: Game;
+    var scene: PlayerScene;
+    var player: Player;
+    var block: Block;
+    var keyDown: EventListenerOrEventListenerObject;
+    var keyUp: EventListenerOrEventListenerObject;
+    var testText: GameObjects.Text;
+
+    function delay(ms) {
+        return new Promise((resolve) => {
+            scene.time.addEvent({
+                delay: ms,
+                callback: resolve,
+            });
+        });
+    }
+
+    function waitForGameEvent(eventName: string) {
+        return new Promise((resolve) => {
+            scene.events.addListener(eventName, () => {
+                resolve();
+            });
+        });
+    }
+
+    before((done) => {
+        const config: Phaser.Types.Core.GameConfig = {
+            title: 'Lava Escape',
+            width: 800,
+            height: 600,
+            scene: [PlayerScene],
+            type: Phaser.AUTO, // TODO set to HEADLESS from env var
+            parent: 'content',
+            callbacks: {
+                postBoot: function (_game) {
+                    game = _game;
+                    scene = game.scene.getScene('MainScene') as PlayerScene;
+                    scene.setOnSceneCreated(() => {
+                        scene.setOnSceneCreated(null);
+                        done();
+                    });
+                },
+            },
+        };
+        new TestGame(config);
+    });
+
+    beforeEach(function () {
+        scene.onTestStart((_player: Player, _block: Block) => {
+            player = _player;
+            block = _block;
+        });
+        window.removeEventListener('keydown', keyDown);
+        window.removeEventListener('keyup', keyUp);
+        scene.game.events.removeAllListeners('grounded');
+        testText = scene.add.text(0, 30, this.currentTest.title, {
```

Comment:
If using `karma-env-preprocessor` to insert env var, also hide test text when headless.

Reply:
Will be dealt with in #39